lib/heuristics/concerns/periodic_end_phase.rb

```diff
@@ -106,52 +65,48 @@ def add_missing_visits
       vehicle_id = best_cost[1][:vehicle]
       log "It is interesting to add #{id} at day #{day} on #{vehicle_id}", level: :debug
 
-      @ids_to_renumber |= [id]
       insert_point_in_route(@candidate_routes[vehicle_id][day], best_cost[1])
-      @output_tool&.add_single_visit(day, @services_data[id][:used_days], id, @services_data[id][:raw].visits_number)
+      @output_tool&.add_single_visit(day, @services_assignment[id][:days], id, @services_data[id][:raw].visits_number)
 
       costs = update_costs(costs, best_cost)
     end
-
-    reaffect_visits_number
   end
 
-  def update_costs(costs, best_cost)
-    # update costs for inserted id, available_days changed
-    uninserted_set = @uninserted.select{ |_key, info| info[:original_id] == best_cost[0] }.keys
-    @uninserted.delete(uninserted_set.first)
-    if uninserted_set.size > 1
-      available_days = days_respecting_lapse(best_cost[0], @candidate_routes[best_cost[1][:vehicle]])
-
-      cost = find_best_day_cost(@candidate_routes[best_cost[1][:vehicle]], best_cost[0], available_days)
-
-      if cost
-        costs[best_cost[0]] = cost
-      else
-        costs.delete(best_cost[0])
-      end
+  # Updates costs after insertion of inserted_cost
+  # Costs of inserted id should be updated : some days are no longer available because of lapses
+  # Costs corresponding to this route should updated : they may not be assignable anymore
+  def update_costs(costs, inserted_cost_data)
+    # Update inserted_id costs
+    inserted_id = inserted_cost_data[0]
+    inserted_cost = inserted_cost_data[1]
+    if @services_assignment[inserted_id][:missing_visits].positive?
+      costs[inserted_id] = find_best_day_cost(@candidate_routes[inserted_cost[:vehicle]], inserted_id)
+      costs.delete(inserted_id) unless costs[inserted_id]
     else
-      costs.delete(best_cost[0])
+      costs.delete(inserted_id)
     end
 
-    # update costs for all ids that should take place at this day. Route changed so cost can change too.
+    # Update costs on this route
     costs.each{ |id, info|
-      next if info[:day] != best_cost[1][:day] && info[:vehicle] != best_cost[1][:vehicle]
-
-      cost = find_best_day_cost(@candidate_routes[info[:vehicle]], id)
+      next unless info[:day] == inserted_cost[:day] &&
```

Comment:
the logical equivalence of !a && !b is !(a || b)
```suggestion
      next unless info[:day] == inserted_cost[:day] ||
```

Reply:
There was a mistake before : it should avec been 

```
info[:day] != best_cost[1][:day] || info[:vehicle] != best_cost[1][:vehicle]
```

We do want to have both same day and vehicle, otherwise it is a different route and it will not be impacted.